### PR TITLE
tls: Account for a change in BoringSSL verify depth semantics

### PR DESCRIPTION
**Pull request by davidben**
Commit Message:
Prior to OpenSSL 1.1.0, the certificate depth limit in OpenSSL omitted the leaf but included the trust anchor. That is, if your chain was Leaf, Intermediate, Root, any depth limit of 2 or more allowed the certificate.

OpenSSL 1.1.0 included d9b8b89bec4480de3a10bdaf9425db371c19145b, which was described as a cleanup change to X509_verify_cert. However, this change the semantics of the depth limit to omit *both* the leaf and trust anchor. So the example above was accepted also at depth limit 1. This is also why common.proto had a comment about different semantics between the libraries.

BoringSSL originally forked a little before 1.0.2, so it had the older OpenSSL behavior. Now that the new behavior has been in OpenSSL upstream for a while, BoringSSL plans to match the new behavior in https://boringssl-review.googlesource.com/c/boringssl/+/64707/

This change makes Envoy compatible with BoringSSLs before and after that change. When BORINGSSL_API_VERSION is new enough, we adjust the value before passing it in, to preserve the original semantics. I'm assuming here that Envoy would prefer to maintain its existing semantics, rather than change the test expectation. I've also removed the comment about backend-specific behavior difference. Supposing Envoy prefers to maintain existing semantics, any OpenSSL port of Envoy should similarly adjust the value on OpenSSL 1.1.0 and up.

Along the way, fix an overflow. maxVerifyDepth is a uint32_t, but the OpenSSL API takes an int. When we exceed INT_MAX, saturate the cast.

Additional Description:
Risk Level: Low (PR is intended to prevent a behavior change)
Testing: Ran default_validator_test on old and new BoringSSL
Docs Changes: Clarified docs for max_verify_depth
Release Notes: N/A
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/transport_sockets/tls/cert_validator/default_validator.cc
Comment (ggreenway) [inline source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:535]:
```suggestion
    uint32_t max_verify_depth = std::min(config_->maxVerifyDepth().value(), uint32_t{INT_MAX});
```

**Review by ggreenway — CHANGES_REQUESTED** [review]
LGTM, modulo one nit

/wait

File: source/extensions/transport_sockets/tls/cert_validator/default_validator.cc
Comment (davidben) [reply source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:535]: That will cause the subtraction below to potentially overflow.

File: source/extensions/transport_sockets/tls/cert_validator/default_validator.cc
Comment (adisuissa) [inline source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:541]: if `max_verify_depth` is 0, won't this overflow?

File: source/extensions/transport_sockets/tls/cert_validator/default_validator.cc
Comment (adisuissa) [inline source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:535]: nit: const

**Review by adisuissa — COMMENTED** [review]
Thanks! left a minor comment.
/lgtm api

File: source/extensions/transport_sockets/tls/cert_validator/default_validator.cc
Comment (davidben) [reply source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:541]: No, that's why I'm setting it to `int`, contrary to the suggestion in. https://github.com/envoyproxy/envoy/pull/31287#discussion_r1424411253

0 - 1 is perfectly well-defined in C, and the API, at least in BoringSSL's implementation, does work with -1. It's just a bit odd, since it means we'll only accept explicitly-trusted self-signed roots (a questionably well-defined concept in X.509). But, if that's what Envoy is requesting, okay whatever. :)

Comment (davidben) [reply source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:541]: I've added a test for it, so we can keep it vaguely working but as this is a pretty incoherent thing to want, I don't think we'd ever officially support this in BoringSSL, no.

Comment (davidben) [reply source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:541]: If you prefer, I can clamp it to 0 instead of -1. It would mean the new BoringSSL will change your behavior ever-so-slightly, but considering that maxVerifyDepth = 0, under your existing semantics, is incoherent, I don't think the behavior change matters much.

Comment (adisuissa) [reply source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:541]: I see. Thanks for providing the additional context.
I'll let @ggreenway to weigh in as a senior-maintainer in this domain.

Comment (ggreenway) [reply source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:541]: I think I'd rather clamp at 0, and take the slight behavior change.

File: source/extensions/transport_sockets/tls/cert_validator/default_validator.cc
Comment (davidben) [reply source/extensions/transport_sockets/tls/cert_validator/default_validator.cc]: n/a with changes

File: source/extensions/transport_sockets/tls/cert_validator/default_validator.cc
Comment (davidben) [reply source/extensions/transport_sockets/tls/cert_validator/default_validator.cc]: Done.

**Review by ggreenway — APPROVED** [review]
Thanks!

File: api/envoy/extensions/transport_sockets/tls/v3/common.proto
Comment (ggreenway) [inline api/envoy/extensions/transport_sockets/tls/v3/common.proto:541]: It seems that the protodoc tool demands double-backticks here

```suggestion
  // This matches the semantics of ``SSL_CTX_set_verify_depth`` in OpenSSL 1.0.x and older versions of BoringSSL. It differs from ``SSL_CTX_set_verify_depth``
```

**Review by ggreenway — CHANGES_REQUESTED** [review]
/wait

**Review by ggreenway — APPROVED** [review]

File: source/extensions/transport_sockets/tls/cert_validator/default_validator.cc
Comment (adisuissa) [inline source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:535]: nit: const

**Review by adisuissa — APPROVED** [review]
Thanks, LGTM!
CI issue seems to be fixed by #31326. Can you please merge head again?

File: source/extensions/transport_sockets/tls/cert_validator/default_validator.cc
Comment (davidben) [reply source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:535]: Already mentioned in another comment thread. The value is not const. See the immediately following line.

**Comment by davidben** [comment]
> Can you please merge head again?

Done.

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
